src/neo/Oracle/OracleService.cs

```diff
@@ -265,24 +471,112 @@ public static Transaction CreateResponseTransaction(StoreView snapshot, OracleEx
                         AllowedContracts = new UInt160[]{ NativeContract.Oracle.Hash },
                         Scopes = WitnessScope.CustomContracts
                     }
-                },
-                NetworkFee = 1_000_000, // TODO: Define fee
-                SystemFee = 1_000_000,  // TODO: Define fee
-                Witnesses = new Witness[0],
-                Script = script.ToArray()
+                }
             };
         }
 
+        #region Sorts
+
+        private static int SortRequest(KeyValuePair<UInt256, RequestItem> a, KeyValuePair<UInt256, RequestItem> b)
+        {
+            return a.Value.CompareTo(b.Value);
+        }
+
+        private static int SortEnqueuedRequest(KeyValuePair<UInt256, Transaction> a, KeyValuePair<UInt256, Transaction> b)
+        {
+            return a.Value.FeePerByte.CompareTo(b.Value.FeePerByte);
+        }
+
+        private static int SortResponse(KeyValuePair<UInt256, ResponseCollection> a, KeyValuePair<UInt256, ResponseCollection> b)
+        {
+            return a.Value.Timestamp.CompareTo(b.Value.Timestamp);
+        }
+
+        #endregion
+
         /// <summary>
-        /// Sign
+        /// Try add oracle response payload
         /// </summary>
-        /// <param name="tx">Transaction</param>
-        /// <returns>Signature</returns>
-        private void Sign(Transaction tx, out byte[] signature)
+        /// <param name="snapshot">Snapshot</param>
+        /// <param name="oracle">Oracle</param>
+        /// <param name="responseTx">Response TX (from OracleService)</param>
+        /// <returns>True if it was added</returns>
+        internal bool TryAddOracleResponse(StoreView snapshot, OraclePayload oracle, Transaction responseTx)
+        {
+            return TryAddOracleResponse(snapshot, new ResponseItem(oracle, responseTx));
+        }
+
+        /// <summary>
+        /// Try add oracle response payload
+        /// </summary>
+        /// <param name="snapshot">Snapshot</param>
+        /// <param name="response">Response</param>
+        /// <returns>True if it was added</returns>
+        private bool TryAddOracleResponse(StoreView snapshot, ResponseItem response)
+        {
+            if (!response.Msg.Verify(snapshot))
+            {
+                return false;
+            }
+
+            // Find the request tx
+
+            if (_pendingOracleRequest.TryGetValue(response.Data.TransactionRequestHash, out var request))
+            {
+                // Append the signature if it's possible
+
+                if (request.AddSinature(response))
+                {
+                    if (request.IsCompleted)
+                    {
+                        // Done! Send to mem pool
```

Comment:
I see multisignature transaction forming there and I'm wondering can we make it a generic mechanism not tied to OracleService? I mean having some protocol to glue together several signatures for regular multisignature transactions that OracleService could then reuse for its purposes.

At the moment we have to sign these transactions via some off-chain mechanism and then when all required signatures are in place push the result to the network, but with this mechanism we could push incomplete (in terms of required signatures) transactions to the network, let the network collect other incomplete transactions and then generate and push the result as soon as the required number of signatures is reached.

This is what already is being done here and if we're to extend it a little for regular transactions I think we can gain some interesting functionality. NeoFS actually could use it then for its purposes (we have a scenario where inner ring nodes need to agree on something and every node can sign for itself, but they need to create this multisignature transaction in the end).

Reply:
It's a very good proposal, but maybe it's hard to solve, now we only considered to do this with oracle tx, but it's something very interesting, could you open a new issue for it?

Reply:
#1573